### ISSUE TITLE
Long-running transactions are not protected from multithreaded access

### ISSUE THREAD

**Issue by bigwad** [Starcounter/Starcounter#4021]
Consider this code:
```c#
using (var tran = new Transaction())
{
  Scheduling.ScheduleTask(() =>
  {
    tran.Scope(() =>
    {
      var t = new TestClass();
     });
   });

  tran.Scope(() =>
  {
    var t = new TestClass();
    tran.Commit();
  });
}
```

With the current implementation of Transaction it's possible to simultaneously work with Transaction on two different context in two parallel threads. It violates kernel requirements to synchronization and will corrupt kernel state in the current process. Nothing in documentation warns the app developer either.
As it corrupts only current process' state - we may just explicitly document Transaction object as non thread-safe, but probably it's worth to prevent such behavior programmatically.

cc @chrhol

**Comment by chrhol** [comment]
Missed this one. 

In v2.2 the check for accessing a long-running transaction on the correct scheduler was done in the kernel, and as I understand this is no longer the case for the current kernel (2.3+).

Just so I understand it correctly: it's ok for the transaction to be used from several schedulers, just not at the same time?

**Comment by bigwad** [comment]
> Just so I understand it correctly: it's ok for the transaction to be used from several schedulers, just not at the same time?

Right.

**Comment by chrhol** [comment]
@dan31: is this serious enough so that we should aim to fix in 2.3 or is it enough with develop?

**Comment by dan31** [comment]
@chrhol I suggest to fix it in 2.3 too. We already know that now returning back the proper checks strikes us badly. Let's consider such fix as critical.

**Comment by chrhol** [comment]
I think I'm quite finished with the needed changes. I'm currently waiting for a personal nightly build to succeed before pushing. 

How it works is that if you're trying to use a (long-running) transaction from more than one thread at the same time, any additional calls to `Scope` for the same transaction will throw an exception (`ScErrTransactionAlreadyBound`) 

To make this work I had to change how transactions are used internally in XSON and I obsoleted a bunch of overloads for `Scope` with different parameters in `ITransaction`, `Transaction`, `JsonExtension`, `ITransactionManager` and `TransactionManager`.  

These overloads were never used (from what I could find) and the customer code I checked and tested works without any additional changes.

**Comment by chrhol** [comment]
Pushed to RC and develop now. I will add a note in the documentation for this change before closing the issue.

**Comment by chrhol** [comment]
Doc for long-running transaction is updated (RC and develop branches).

**Comment by chrhol** [comment]
Reopening this issue since it seems I introduced some nasty bugs with these changes that I haven't been succesful in fixing. I have reverted the changes for now in both `v2.3` and `develop` branches.

**Comment by chrhol** [comment]
Reapplied the fix previously reverted including the missing piece, a missing check for an invalid index in one place. How this could have worked without this check, I have no idea. But it did and the only place it could be reproduced was on a machine without VS2015 installed (it actually started working when installing VS) so it took some time to find.

Anyway the working fix is pushed to v2.3 and develop now.
